File: tests/output/dart/test/manual/rate_limit_retry_test.dart
import 'package:algolia_client_core/algolia_client_core.dart';
import 'package:algolia_client_core/src/transport/rate_limit.dart';
import 'package:test/test.dart';

const ok = HttpResponse(200, {'message': 'ok'});
const serverError = AlgoliaApiException(500, {'message': 'server error'});
const rateLimitedBody = {'message': 'Too many requests'};

AlgoliaApiException rateLimited({
  Map<String, String>? headers,
  String? correlationId,
}) =>
    AlgoliaApiException(
      429,
      rateLimitedBody,
      correlationId: correlationId,
      headers: headers,
    );

/// Answers each request from a per-host script and records every call. A
/// scripted [AlgoliaException] is thrown, an [HttpResponse] is returned as is
/// (a requester may report errors either way); the last outcome repeats.
final class ScriptedRequester implements Requester {
  final Map<String, List<Object>> scripts;
  final List<String> calls = [];

  ScriptedRequester(this.scripts);

  @override
  Future<HttpResponse> perform(HttpRequest request) async {
    final hostUrl = request.host.url;
    calls.add(hostUrl);
    final script = scripts[hostUrl]!;
    final outcome = script.length > 1 ? script.removeAt(0) : script.first;
    if (outcome is AlgoliaException) throw outcome;
    return outcome as HttpResponse;
  }

  @override
  Duration? connectTimeout;

  @override
  void setConnectTimeout(Duration connectTimeout) {
    this.connectTimeout = connectTimeout;
  }

  @override
  void setClientApiKey(String apiKey) {}

  @override
  void close() {}
}

RetryStrategy strategy(
  ScriptedRequester requester,
  List<Duration> waits, {
  List<String> hosts = const ['a'],
  int? maxRateLimitRetries,
}) =>
    RetryStrategy(
      requester: requester,
      readTimeout: const Duration(seconds: 5),
      writeTimeout: const Duration(seconds: 30),
      hosts: hosts.map((url) => Host(url: url)),
      maxRateLimitRetries: maxRateLimitRetries,
      sleep: (duration) async {
        waits.add(duration);
      },
    );

const getRequest = ApiRequest(method: RequestMethod.get, path: '/1/test');

final isRateLimitedException = isA<AlgoliaApiException>()
    .having((e) => e.statusCode, 'statusCode', 429)
    .having((e) => e.error, 'error', rateLimitedBody);

void main() {
  group('parseRetryAfter', () {
    const oneSecond = Duration(seconds: 1);
    const ceiling = Duration(seconds: maxRateLimitWaitSeconds);

    test('honours a positive whole number of seconds', () {
      expect(parseRetryAfter({'retry-after': '2'}), const Duration(seconds: 2));
      expect(
          parseRetryAfter({'retry-after': ' 3 '}), const Duration(seconds: 3));
      expect(
          parseRetryAfter({'retry-after': '007'}), const Duration(seconds: 7));
    });

    test('matches the header name case-insensitively', () {
      expect(parseRetryAfter({'Retry-After': '4'}), const Duration(seconds: 4));
      expect(parseRetryAfter({'RETRY-AFTER': '5'}), const Duration(seconds: 5));
    });

    test('falls back to one second when the header is missing or unusable', () {
      expect(parseRetryAfter(null), oneSecond);
      expect(parseRetryAfter({}), oneSecond);
      expect(parseRetryAfter({'x-other': '2'}), oneSecond);
      expect(parseRetryAfter({'retry-after': ''}), oneSecond);
      expect(parseRetryAfter({'retry-after': '0'}), oneSecond);
      expect(parseRetryAfter({'retry-after': '-5'}), oneSecond);
      expect(parseRetryAfter({'retry-after': 'abc'}), oneSecond);
      expect(parseRetryAfter({'retry-after': '1.5'}), oneSecond);
      expect(
        parseRetryAfter({'retry-after': 'Wed, 21 Oct 2015 07:28:00 GMT'}),
        oneSecond,
      );
    });

    test('accepts ASCII digits only', () {
      // Arabic-Indic three.
      expect(parseRetryAfter({'retry-after': '٣'}), oneSecond);
    });

    test('waits the Duration ceiling for values it cannot hold', () {
      expect(ceiling.isNegative, isFalse);
      expect(parseRetryAfter({'retry-after': '9223372036854'}), ceiling);
      expect(parseRetryAfter({'retry-after': '9223372036855'}), ceiling);
      expect(
        parseRetryAfter({'retry-after': '99999999999999999999'}),
        ceiling,
      );
    });
  });

  group('resolveMaxRateLimitRetries', () {
    test('null means the default, negative means zero', () {
      expect(resolveMaxRateLimitRetries(null), defaultMaxRateLimitRetries);
      expect(resolveMaxRateLimitRetries(null), 3);
      expect(resolveMaxRateLimitRetries(0), 0);
      expect(resolveMaxRateLimitRetries(-1), 0);
      expect(resolveMaxRateLimitRetries(5), 5);
    });
  });

  group('RetryStrategy', () {
    test('retries a 429 on the same host after Retry-After', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [
          rateLimited(headers: {'retry-after': '2'}),
          ok
        ],
      });
      final retryStrategy = strategy(requester, waits);

      final response = await retryStrategy.execute(request: getRequest);

      expect(response, {'message': 'ok'});
      expect(requester.calls, ['a', 'a']);
      expect(waits, [const Duration(seconds: 2)]);
      final host = retryStrategy.hosts.single;
      expect(host.isUp, isTrue);
      expect(host.retryCount, 0);
    });

    test('waits one second when Retry-After is missing', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [rateLimited(), ok],
      });

      await strategy(requester, waits).execute(request: getRequest);

      expect(requester.calls, ['a', 'a']);
      expect(waits, [const Duration(seconds: 1)]);
    });

    test('also retries when the requester returns the 429 as a response',
        () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [
          const HttpResponse(429, rateLimitedBody,
              headers: {'Retry-After': '1'}),
          ok,
        ],
      });

      await strategy(requester, waits).execute(request: getRequest);

      expect(requester.calls, ['a', 'a']);
      expect(waits, [const Duration(seconds: 1)]);
    });

    test('fails on the first 429 when maxRateLimitRetries is 0', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [
          rateLimited(headers: {'retry-after': '2'})
        ],
      });
      final retryStrategy = strategy(requester, waits, maxRateLimitRetries: 0);

      await expectLater(
        retryStrategy.execute(request: getRequest),
        throwsA(isRateLimitedException),
      );
      expect(requester.calls, ['a']);
      expect(waits, isEmpty);
    });

    test('treats a negative maxRateLimitRetries like 0', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [rateLimited()],
      });
      final retryStrategy = strategy(requester, waits, maxRateLimitRetries: -1);

      await expectLater(
        retryStrategy.execute(request: getRequest),
        throwsA(isRateLimitedException),
      );
      expect(requester.calls, ['a']);
      expect(waits, isEmpty);
    });

    test('surfaces the 429 once the default budget is used up', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [
          rateLimited(headers: {'retry-after': '2'})
        ],
      });
      final retryStrategy = strategy(requester, waits);

      await expectLater(
        retryStrategy.execute(request: getRequest),
        throwsA(isRateLimitedException),
      );
      expect(requester.calls, ['a', 'a', 'a', 'a']);
      expect(waits, List.filled(3, const Duration(seconds: 2)));
      expect(retryStrategy.hosts.single.isUp, isTrue);
    });

    test('restores the connect timeout before waiting on a 429', () async {
      final requester = ScriptedRequester({
        'a': [
          rateLimited(headers: {'retry-after': '2'}),
          ok
        ],
      });
      requester.setConnectTimeout(const Duration(seconds: 2));
      final seenWhileWaiting = <Duration?>[];
      final retryStrategy = RetryStrategy(
        requester: requester,
        readTimeout: const Duration(seconds: 5),
        writeTimeout: const Duration(seconds: 30),
        hosts: [Host(url: 'a')],
        sleep: (_) async {
          seenWhileWaiting.add(requester.connectTimeout);
        },
      );

      await retryStrategy.execute(
        request: getRequest,
        options: const RequestOptions(connectTimeout: Duration(seconds: 9)),
      );

      // The per-call override must not leak to concurrent requests for the
      // length of the wait, so it is restored before sleep runs.
      expect(seenWhileWaiting, [const Duration(seconds: 2)]);
      expect(requester.connectTimeout, const Duration(seconds: 2));
    });

    test('still fails over to the next host on a 5xx', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [serverError],
        'b': [ok],
      });
      final retryStrategy = strategy(requester, waits, hosts: ['a', 'b']);

      await retryStrategy.execute(request: getRequest);

      expect(requester.calls, ['a', 'b']);
      expect(waits, isEmpty);
      expect(retryStrategy.hosts.first.isUp, isFalse);
      expect(retryStrategy.hosts.last.isUp, isTrue);
    });

    test('shares the budget across hosts', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [rateLimited(), serverError],
        'b': [rateLimited()],
      });
      final retryStrategy = strategy(requester, waits, hosts: ['a', 'b']);

      await expectLater(
        retryStrategy.execute(request: getRequest),
        throwsA(isRateLimitedException),
      );
      expect(requester.calls, ['a', 'a', 'b', 'b', 'b']);
      expect(waits, hasLength(3));
    });

    test('keeps the waited-out 429 visible when every host fails', () async {
      final waits = <Duration>[];
      final requester = ScriptedRequester({
        'a': [rateLimited(correlationId: 'corr-429'), serverError],
        'b': [serverError],
      });
      final retryStrategy = strategy(requester, waits, hosts: ['a', 'b']);

      await expectLater(
        retryStrategy.execute(request: getRequest),
        throwsA(
          isA<UnreachableHostsException>().having(
            (e) => e.errors,
            'errors',
            [
              isRateLimitedException.having(
                  (e) => e.correlationId, 'correlationId', 'corr-429'),
              serverError,
              serverError,
            ],
          ),
        ),
      );
      expect(waits, [const Duration(seconds: 1)]);
    });

    test('create resolves maxRateLimitRetries from ClientOptions', () {
      final requester = ScriptedRequester({});
      RetryStrategy create(int? maxRateLimitRetries) => RetryStrategy.create(
            segment: AgentSegment(value: 'Test', version: '0.0.0'),
            appId: 'appId',
            apiKey: 'apiKey',
            defaultHosts: () => [Host(url: 'a')],
            options: ClientOptions(
              requester: requester,
              maxRateLimitRetries: maxRateLimitRetries,
            ),
          );

      expect(create(null).maxRateLimitRetries, 3);
      expect(create(0).maxRateLimitRetries, 0);
      expect(create(-2).maxRateLimitRetries, 0);
      expect(create(7).maxRateLimitRetries, 7);
    });
  });

  group('ClientOptions', () {
    test('includes maxRateLimitRetries in equality, hashCode and toString', () {
      const zero = ClientOptions(maxRateLimitRetries: 0);
      const alsoZero = ClientOptions(maxRateLimitRetries: 0);
      const unset = ClientOptions();

      expect(zero, alsoZero);
      expect(zero.hashCode, alsoZero.hashCode);
      expect(zero, isNot(unset));
      expect(zero.toString(), contains('maxRateLimitRetries: 0'));
      expect(unset.toString(), contains('maxRateLimitRetries: null'));
    });
  });
}

File: tests/output/ruby/test/manual/rate_limit_retry_test.rb
# frozen_string_literal: true

require "algolia"
require "json"
require "test/unit"

# Requester answering each host from its own queue of scripted responses, in
# order, repeating the last one once the queue is spent. Records every attempt.
class ScriptedRequester
  Attempt = Struct.new(:host, :connect_timeout)

  attr_reader :attempts

  # @param script [Hash{String => Array<Hash>}] Response options per host url
  def initialize(script)
    @script = script.transform_values(&:dup)
    @attempts = []
  end

  def send_request(host, _method, _path, _body, _query_params, _headers, _timeout, connect_timeout)
    @attempts << Attempt.new(host, connect_timeout)
    queue = @script.fetch(host.url)
    opts = queue.length > 1 ? queue.shift : queue.first

    Algolia::Http::Response.new({host: host}.merge(opts))
  end
end

class TestRateLimitRetry < Test::Unit::TestCase
  include CallType

  OK = {status: 200, body: JSON.generate({:message => "ok"}), headers: {}}.freeze
  SERVER_ERROR = {status: 500, error: "unavailable", headers: {}}.freeze

  def rate_limited(headers = {"Retry-After" => "2"})
    {status: 429, error: JSON.generate({:message => "Too many requests"}), headers: headers}
  end

  def html_rate_limited
    {
      status: 429,
      error: "<html><body>429 Too Many Requests</body></html>",
      headers: {"content-type" => "text/html; charset=utf-8"},
      reason_phrase: "Too Many Requests"
    }
  end

  def host(url)
    Algolia::Transport::StatefulHost.new(url, accept: READ | WRITE)
  end

  # Builds a search client over the scripted requester, recording every wait
  # instead of sleeping.
  def client_for(script, opts = {})
    @requester = ScriptedRequester.new(script)
    @waits = []
    @hosts = script.keys.map { |url| host(url) }
    config = Algolia::Configuration.new(
      "test-app-id",
      "test-api-key",
      @hosts,
      "Search",
      {requester: @requester}.merge(opts)
    )
    client = Algolia::SearchClient.create_with_config(config)
    client.api_client.transporter.sleeper = ->(seconds) { @waits << seconds }
    client
  end

  def wait_seconds(headers)
    client = client_for({"localhost" => [OK]})
    client.api_client.transporter.send(:rate_limit_wait_seconds, headers)
  end

  def test_parses_retry_after_whole_seconds
    assert_equal(2, wait_seconds({"Retry-After" => "2"}))
    assert_equal(3, wait_seconds({"Retry-After" => " 3 "}))
    assert_equal(7, wait_seconds({"Retry-After" => "007"}))
  end

  def test_retry_after_header_key_is_case_insensitive
    assert_equal(5, wait_seconds({"retry-after" => "5"}))
    assert_equal(5, wait_seconds({"RETRY-AFTER" => "5"}))
    assert_equal(5, wait_seconds({:"Retry-After" => "5"}))
  end

  def test_invalid_retry_after_waits_one_second
    assert_equal(1, wait_seconds({}))
    assert_equal(1, wait_seconds({"Retry-After" => ""}))
    assert_equal(1, wait_seconds({"Retry-After" => "0"}))
    assert_equal(1, wait_seconds({"Retry-After" => "-5"}))
    assert_equal(1, wait_seconds({"Retry-After" => "abc"}))
    assert_equal(1, wait_seconds({"Retry-After" => "1.5"}))
    assert_equal(1, wait_seconds({"Retry-After" => "Wed, 21 Oct 2015 07:28:00 GMT"}))
    # Non-ASCII digits are not a whole number of seconds.
    assert_equal(1, wait_seconds({"Retry-After" => "٣"}))
  end

  def test_missing_headers_wait_one_second
    # Http::Response defaults headers to "" on timeouts and network failures.
    assert_equal(1, wait_seconds(""))
    assert_equal(1, wait_seconds(nil))
  end

  def test_huge_retry_after_saturates_at_the_sleep_ceiling
    ceiling = Algolia::Defaults::MAX_RATE_LIMIT_WAIT

    assert_equal(ceiling, wait_seconds({"Retry-After" => (ceiling + 1).to_s}))
    assert_equal(ceiling, wait_seconds({"Retry-After" => "9" * 40}))
    assert_equal(ceiling - 1, wait_seconds({"Retry-After" => (ceiling - 1).to_s}))
  end

  def test_retries_429_on_the_same_host_after_retry_after
    client = client_for({"host-a" => [rate_limited, OK]})

    response = client.custom_get_with_http_info("1/test")

    assert_equal(200, response.status)
    assert_equal([2], @waits)
    assert_equal(%w[host-a host-a], @requester.attempts.map { |attempt| attempt.host.url })
    # A waited-out 429 is not a failover: the host stays up and its connect
    # timeout is not inflated on the second attempt.
    assert_true(@hosts.first.up)
    assert_equal(0, @hosts.first.retry_count)
    assert_equal(@requester.attempts.first.connect_timeout, @requester.attempts.last.connect_timeout)
  end

  def test_retries_429_with_a_one_second_wait_when_retry_after_is_missing
    client = client_for({"host-a" => [rate_limited({}), OK]})

    response = client.custom_get_with_http_info("1/test")

    assert_equal(200, response.status)
    assert_equal([1], @waits)
    assert_equal(2, @requester.attempts.length)
  end

  def test_zero_retries_fails_on_the_first_429_without_waiting
    client = client_for({"host-a" => [rate_limited, OK]}, max_rate_limit_retries: 0)

    error = assert_raise(Algolia::AlgoliaHttpError) { client.custom_get("1/test") }

    assert_equal(429, error.code)
    assert_equal("429: Too many requests", error.message)
    assert_equal([], @waits)
    assert_equal(1, @requester.attempts.length)
  end

  def test_negative_retries_behave_like_zero
    client = client_for({"host-a" => [rate_limited, OK]}, max_rate_limit_retries: -1)

    error = assert_raise(Algolia::AlgoliaHttpError) { client.custom_get("1/test") }

    assert_equal(429, error.code)
    assert_equal([], @waits)
    assert_equal(1, @requester.attempts.length)
  end

  def test_returns_429_after_the_budget_is_used_up
    client = client_for({"host-a" => [rate_limited]})

    error = assert_raise(Algolia::AlgoliaHttpError) { client.custom_get("1/test") }

    assert_equal(429, error.code)
    assert_equal("429: Too many requests", error.message)
    assert_equal([2, 2, 2], @waits)
    assert_equal(4, @requester.attempts.length)
    assert_true(@hosts.first.up)
  end

  def test_html_429_is_raised_with_its_reason_phrase_after_the_budget
    client = client_for({"host-a" => [html_rate_limited]})

    error = assert_raise(Algolia::AlgoliaHttpError) { client.custom_get("1/test") }

    assert_equal("429: Too Many Requests", error.message)
    assert_equal([1, 1, 1], @waits)
    assert_equal(4, @requester.attempts.length)
  end

  def test_5xx_still_fails_over_to_the_next_host
    client = client_for({"host-a" => [SERVER_ERROR], "host-b" => [OK]})

    response = client.custom_get_with_http_info("1/test")

    assert_equal(200, response.status)
    assert_equal([], @waits)
    assert_equal(%w[host-a host-b], @requester.attempts.map { |attempt| attempt.host.url })
    assert_false(@hosts.first.up)
  end

  def test_budget_is_shared_across_hosts
    client = client_for(
      {
        "host-a" => [rate_limited, SERVER_ERROR],
        "host-b" => [rate_limited]
      }
    )

    error = assert_raise(Algolia::AlgoliaHttpError) { client.custom_get("1/test") }

    assert_equal(429, error.code)
    assert_equal([2, 2, 2], @waits)
    assert_equal(%w[host-a host-a host-b host-b host-b], @requester.attempts.map { |attempt| attempt.host.url })
  end

  def test_waited_out_429_stays_visible_when_every_host_fails
    client = client_for(
      {
        "host-a" => [rate_limited({"Retry-After" => "2", "Correlation-ID" => "abc"}), SERVER_ERROR],
        "host-b" => [SERVER_ERROR]
      }
    )

    error = assert_raise(Algolia::AlgoliaUnreachableHostError) { client.custom_get("1/test") }

    assert_equal([2], @waits)
    assert_equal("abc", error.correlation_id)
    assert_true(error.message.end_with?("(Correlation-ID: abc)"))
    assert_equal(
      [
        {host: "host-a", error: JSON.generate({:message => "Too many requests"})},
        {host: "host-a", error: "unavailable"},
        {host: "host-b", error: "unavailable"}
      ],
      error.errors
    )
  end

  def test_configuration_defaults_to_three_retries
    config = Algolia::Configuration.new("test-app-id", "test-api-key", [host("localhost")], "Search")

    assert_nil(config.max_rate_limit_retries)
    assert_equal(3, Algolia::Defaults::MAX_RATE_LIMIT_RETRIES)
  end

  def test_create_forwards_max_rate_limit_retries
    client = Algolia::SearchClient.create("test-app-id", "test-api-key", {max_rate_limit_retries: 0})

    assert_equal(0, client.api_client.config.max_rate_limit_retries)
  end

  def test_transformation_options_override_reaches_the_ingestion_transporter
    client = Algolia::SearchClient.with_transformation(
      "test-app-id",
      "test-api-key",
      Algolia::TransformationOptions.new("us", max_rate_limit_retries: 0)
    )

    ingestion_config = client.instance_variable_get(:@ingestion_transporter).api_client.config
    assert_equal(0, ingestion_config.max_rate_limit_retries)
  end

  def test_ingestion_transporter_does_not_inherit_the_parent_setting
    client = Algolia::SearchClient.with_transformation(
      "test-app-id",
      "test-api-key",
      Algolia::TransformationOptions.new("us"),
      {max_rate_limit_retries: 0}
    )

    assert_equal(0, client.api_client.config.max_rate_limit_retries)
    ingestion_config = client.instance_variable_get(:@ingestion_transporter).api_client.config
    # nil resolves to the Ingestion default of 3 in the transporter.
    assert_nil(ingestion_config.max_rate_limit_retries)
  end
end
